### ISSUE TITLE
crypto/tls: no support for SSLv2 handshake

### ISSUE THREAD

**Issue by gopherbot** [golang/go#3930]
by **ian.ragsdale**:

<pre>To reproduce, run the attached TLS echo server and try to connect using the openssl
s_client utility:

&gt; openssl s_client -quiet -connect localhost:50000                                   
                                                      
95052:error:140773F2:SSL routines:SSL23_GET_SERVER_HELLO:sslv3 alert unexpected
message:/SourceCache/OpenSSL098/OpenSSL098-47/src/ssl/s23_clnt.c:602:
&gt; openssl s_client -quiet -connect localhost:50000 -ssl3                             
                                                      
depth=0
/CN=relay.cloudsmtp.com/O=Datran/OU=CapThought/ST=Texas/C=US/L=Austin/emailAddress=ian.ragsdale@gmail.com
verify error:num=20:unable to get local issuer certificate
verify return:1
depth=0
/CN=relay.cloudsmtp.com/O=Datran/OU=CapThought/ST=Texas/C=US/L=Austin/emailAddress=ian.ragsdale@gmail.com
verify error:num=21:unable to verify the first certificate
verify return:1
hi 
&gt;&gt; hi

Note that the s_client is unable to connect with its default settings, but can connect
by specifying a version.

This is because TLS-capable clients that also support connecting to SSLv2 servers start
off by sending a SSLv2 formatted CLIENT-HELLO packet instead of a TLS ClientHello.  The
crypto/tls package does not recognize these packets and so terminates the handshake,
making it impossible to use these clients. However, these ARE compliant TLS clients, and
they set the TLS version correctly in the CLIENT-HELLO packet in order to signify that
they support TLS.  This makes it possible for the crypto/tls code to read these packets,
recognize TLS compliance, and send back a TLS-compliant ServerHello packet, at which
point the entire rest of the conversation is exactly like any other TLS session.

So, in order to support these clients, the only thing that needs to be done is to
recognize a SSLv2 compliant packet correctly and begin a standard TLS conversation. 
Since it seemed fairly simple to do so, I've attached a fairly simple patch to allow
communicating with these clients.  I'm sure it's unacceptable in its current form, but
it functions for me and illustrates the changes that are needed to support this
functionality.

I completely understand (and agree with) the aversion to supporting older versions of
SSL, but this is actually TLS behavior documented in at least the 1.0 and 1.1 RFCs, and
is a fairly minor change, which should greatly improve interop.  Lots of common clients
(recent versions of Java and Ruby at least) do this by default. I can easily configure
my own clients to not do this, but it isn't easy for a lot of people.</pre>

Attachments:

1. <a href="https://storage.googleapis.com/go-attachment/3930/0/tls_echo.go">tls_echo.go</a> (1002 bytes)
1. <a href="https://storage.googleapis.com/go-attachment/3930/0/0001-Allow-SSLv2-compatible-client-hello-so-SSLv2-compati.patch">0001-Allow-SSLv2-compatible-client-hello-so-SSLv2-compati.patch</a> (5479 bytes)

**Comment by elimisteve** [comment]
I've tweaked `erwan.legrand`'s Go 1.1 patch to support Go 1.4.  I applied it to my std lib, and the dreaded failed requests and `tls: unsupported SSLv2 handshake received` error messages appear to be gone!  (See below.)

Having a 95% solution in Go was painful, and after trying many other options (e.g., re-implementing much of my proxy app in another language with a more mature TLS library, using a not-programmable-enough Python proxy, and almost starting from scratch and using Nginx + Lua), this finally appears to be a fully working solution -- and 100% in Go!

To reiterate previous security warnings:

> Before you use this, please read Adam Langley's comments above: the Go TLS stack is not ready for production. For all we know, it might leak your private key. You have been
warned.

To download and use this patch, then recompile the Go standard library on Linux, run something like...

```bash
cd ~
wget https://gist.githubusercontent.com/elimisteve/e3fea232c11fa08146d6/raw/e2457232b63a8f0433326874621565475cdb2524/Allow-SSLv2-ClientHello-go1.4.patch
cd go/src  # cd to wherever your Go source is located
patch -p0 < ~/Allow-SSLv2-ClientHello-go1.4.patch
./all.bash  # recompile
```

Thank you, all!

**Comment by elimisteve** [comment]
For some reason I kept accidently using an old Go runtime after recompiling.  If you've used the above patch and recompiled, and you want to make sure your `go` binary is patched, you can run

    grep 'tls: unsupported SSLv2 handshake received' `which go`

to see if this SSLv2 error from the standard library, which should NOT be in your patched `go`, is present nonetheless.  And if you run `go build myprogram.go`, afterward you can run `grep 'tls: unsupported SSLv2 handshake received' myprogram` to make sure that the Go runtime that has been linked into your `myprogram` binary is the patched one.

**Comment by athornton** [comment]
I just got bitten by this one.

The problem (at least for me) is less ancient clients that actually support SSLv2, and more: the Java 6 default behavior, when making an SSL connection, is to send SSLv2 Hello.

Now, sure, Java 6 is also pretty old, but it's in really wide use in the wild, especially in the context of big horrible Enterprisey apps (which is where this got me).  And in many cases it's not going to be feasible to modify the Java properties of the connecting applications to only do TLS.

**Comment by FiloSottile** [comment]
I rebased on 1.5.2/fixed/polished/optimized that patch a bit and successfully ran it at pretty high load. **It's here: [crypto/tls: support SSLv2 compatibility handshakes](https://gist.github.com/FiloSottile/8e0fe1a3523126e8c581)**. I wouldn't use it if I were you, but if you're desperate, it might not blow up.

**Comment by nathany** [comment]
> I'm tempted to close this bug without action. @agl?

That sounds like a good idea.
https://drownattack.com/

**Comment by bradfitz** [comment]
Hah, indeed. I've wanted to close this bug for a long time, but that's a great excuse.

**Comment by FiloSottile** [comment]
I'm not against closing this, but DROWN is not relevant.

This ticket is about supporting an unfortunate compatibility trick, not the entire protocol, so it would not introduce the vulnerability.

**Comment by bradfitz** [comment]
My thought process was that if people finally kill off SSLv2, this compatibility feature is no longer relevant.

**Comment by rsc** [comment]
I see how killing off SSLv2 servers is technically separate from then
eliminating the SSLv2 compatibility handshake from TLS clients. That said,
Adam's comment above says, "IE on XP was a big source of this and we've
made that mostly moot by disabling SSLv3 by default on trunk. You can also
get the same thing with s_client if you don't update OpenSSL, which is
similarly a bad idea."

Clearly people still run into this. How? What clients still do this?

**Comment by cespare** [comment]
If more anecdata helps, I ran into this within the last couple of months with a third-party service doing HTTPS requests to our servers from Java 6 code (as mentioned elsewhere in this thread, the Java 6 SSL stack will do an SSLv2 Client Hello by default).

**Comment by elimisteve** [comment]
I, too, ran into this SSLv2 issue from a Java client, which was trying to connecting to an HTTPS proxy I wrote in Go.

**Comment by athornton** [comment]
Yeah.  As far as I know, the primary place this shows up is The Very Important Service That Only Runs On Java 6 And No One Understands It Enough To Rewrite It And There Is No Way We Are Going To Crack It Open To Change The SSL Settings Are You Kidding?  Which, on my more cynical days, I think describes about 90% of Enterprise Software.

**Comment by bradfitz** [comment]
Okay, I can reopen this. I was being too hopeful.

@FiloSottile, the tree is open. Can you formally send https://gist.github.com/FiloSottile/8e0fe1a3523126e8c581? It doesn't look too invasive. We can guard it behind a flag (make it opt-in) for more paranoia.

**Comment by FiloSottile** [comment]
Done https://go-review.googlesource.com/20094

Considering that ian.ragsdale, erwan.legrand and @elimisteve contributed to earlier versions of the patch, should we chase their CLAs?

**Comment by gopherbot** [comment]
CL https://golang.org/cl/20094 mentions this issue.

**Comment by rsc** [comment]
I looked at the sequence of patches. The one in the Gist has only trivial changes from the original by Ian Ragsdale, so I think the only CLA we absolutely need is his. 

@iragsdale, could you complete a contributor CLA (https://golang.org/doc/contribute#copyright) so that we can put the SSLv2 compatibility code into Go's standard crypto/tls?

If @ErwanLegrand and @elimisteve want to complete CLAs as well I'm happy to credit them too, but the followup changes appear to have been quite minor.

Thanks everyone.

**Comment by bradfitz** [comment]
@agl replied on https://golang.org/cl/20094 :

> I don't believe that we should land this. It goes against the general trend of dropping support for old SSL versions which keep causing problems.

**Comment by rsc** [comment]
I am glad that we have established that Java 6 is the holdout. I looked around on the internet and found what appears to be [Oracle's official advice on SSL settings as of January 2015](http://www.oracle.com/technetwork/java/javase/documentation/cve-2014-3566-2342133.html).

Under the text "For the following use cases Oracle recommends", it says that _client-side_ uses should set the protocols to exclude the SSLv2 hello, for example:

    sslSocket.setEnabledProtocols(new String[] {"TLSv1"});
    sslEngine.setEnabledProtocols(new String[] {"TLSv1"});

So the Java clients that are still sending SSLv2Hello are not following Oracle's advice.

Based on @agl's Go server-side advice and Oracle's Java client-side advice, I think it is reasonable for Go not to support SSLv2 compatibility handshakes for Java clients.

@iragsdale, @ErwanLegrand, @elimisteve, sorry for the false alarm; no CLA needed after all.

Thanks everyone.
